### PR TITLE
Users app tests

### PR DESCRIPTION
**Pull request by reinbach**
Looking to increase test coverage of users app

### CODE REVIEW

File: EquiTrack/users/tests/test_tasks.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+        self.mapper.countries = {code: country, "UAT": country}
+        profile = ProfileFactory(country=None)
+        self.assertIsNone(profile.country)
+        self.assertFalse(profile.countries_available.count())
+        res = self.mapper._set_special_attr(profile, "country", code)
+        self.assertTrue(res)
+        self.assertEqual(profile.country, country)
+        self.assertTrue(profile.countries_available.count())
+
+    def test_set_attribute_char(self):
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_tasks.py]: docstring?

File: EquiTrack/users/tests/test_tasks.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+        self.assertTrue(profile.countries_available.count())
+
+    def test_create_or_update_user_missing_fields(self):
+        """If missing field, then don't create user record'"""
+        email = "tester@example.com"
+        res = self.mapper.create_or_update_user({"internetaddress": email})
+        self.assertIsNone(res)
+        self.assertFalse(User.objects.filter(email=email).exists())
+
+    def test_create_or_update_user_created(self):
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_tasks.py]: docstring?

File: EquiTrack/users/tests/test_tasks.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        })
+        self.assertIsNone(res)
+        self.assertTrue(User.objects.filter(email=email).exists())
+        self.assertTrue(
+            UserProfile.objects.filter(user__email=email).exists()
+        )
+        user = User.objects.get(email=email)
+        self.assertIn(self.group, user.groups.all())
+
+    def test_create_or_update_user_exists(self):
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_tasks.py]: docstring?

File: EquiTrack/users/tests/test_tasks.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+        self.assertIsNone(profile.supervisor)
+        self.assertTrue(self.mapper._set_supervisor(profile, manager_id))
+        self.assertEqual(profile.supervisor, supervisor.user)
+
+    def test_map_users_no_sections(self):
+        self.assertEqual(self.mapper.section_users, {})
+        self.mapper.map_users()
+        self.assertEqual(self.mapper.section_users, {})
+
+    def test_map_users_response_empty(self):
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_tasks.py]: I had to read this method and the one below twice to find the one line of code where they differ. Docstring please! 😃

File: EquiTrack/users/tests/test_views.py [hunk 5]
```diff
@@ -1,12 +1,66 @@
 from __future__ import unicode_literals
 
 import json
+
+from django.core.urlresolvers import reverse
+from rest_framework import status
+from tenant_schemas.test.client import TenantClient
 from unittest import skip
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_views.py]: `unittest` should be up with the other Python imports

File: EquiTrack/users/tests/test_views.py [hunk 6]
```diff
@@ -251,3 +314,103 @@ def test_minimal_verbosity(self):
...
+
+class TestProfileEdit(FastTenantTestCase):
+    def setUp(self):
+        super(TestProfileEdit, self).setUp()
+        self.client = TenantClient(self.tenant)
+
+    def test_get_non_staff(self):
+        user = UserFactory()
+        self.client.force_login(user)
+        response = self.client.get("/users/profile_view/")
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_views.py]: Possible to use `reverse()` here to get the URL?

File: EquiTrack/users/tests/test_views.py [hunk 7]
```diff
@@ -251,3 +314,103 @@ def test_minimal_verbosity(self):
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertTemplateUsed(response, "users/profile.html")
+        profile = UserProfile.objects.get(user=user)
+        self.assertEqual(profile.office.pk, office.pk)
+        self.assertEqual(profile.section, section)
+        self.assertEqual(profile.job_title, "New Job")
+        self.assertEqual(profile.phone_number, "123-546-7890")
+
+
+# class TestGroupViewSet(APITenantTestCase):
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_views.py]: This should probably go.

File: EquiTrack/users/tasks.py [hunk 8]
```diff
@@ -67,8 +67,8 @@ def _get_country(self, business_area_code):
         return self.countries[business_area_code] or self.countries['UAT']
 
     def _get_section(self, section_name, section_code):
-        if not self.sections[section_name]:
-            self.sections[section_name] = Section.objects.get_or_create(name=section_name, code=section_code)
+        if not self.sections.get(section_name, False):
```
[the remark this answers is not in the document]
Comment (reinbach) [reply EquiTrack/users/tasks.py]: I like to be explicit :)

File: EquiTrack/users/tests/test_models.py [hunk 9]
```diff
@@ -19,3 +107,49 @@ def test_conversion_to_string(self):
...
+        profile = ProfileFactory()
+        profile.partner_staff_member = profile.user.pk
+        profile.save()
+        models.delete_partner_relationship(None, profile.user)
+        profile_updated = models.UserProfile.objects.get(pk=profile.pk)
+        self.assertIsNone(profile_updated.partner_staff_member)
+        user = models.User.objects.get(pk=profile.user.pk)
+        self.assertFalse(user.is_active)
+
+    def test_delete_exception(self):
```
[the remark this answers is not in the document]
Comment (reinbach) [reply EquiTrack/users/tests/test_models.py]: 👍 This was a weak test, improved it a little more

File: EquiTrack/users/tests/test_tasks.py [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+        self.assertIsNone(res)
+        self.assertTrue(User.objects.filter(email=email).exists())
+        self.assertTrue(
+            UserProfile.objects.filter(user__email=email).exists()
+        )
+        user = User.objects.get(email=email)
+        self.assertIn(self.group, user.groups.all())
+
+    def test_create_or_update_user_exists(self):
+        """Ensure gracefull handling if user already exists"""
```
Comment (caktus-philip) [inline EquiTrack/users/tests/test_tasks.py]: `gracefull` == > `graceful`

File: EquiTrack/users/tests/test_tasks.py [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import json
+
+from django.conf import settings
+from mock import patch, Mock
+from tenant_schemas.utils import schema_context
+from unittest import skip
```
[the remark this answers is not in the document]
Comment (reinbach) [reply EquiTrack/users/tests/test_tasks.py]: How come Django is separated from other 3rd party imports?

**Pull request merged**